src/lib/server/auth/passkey.ts

```diff
@@ -0,0 +1,241 @@
+
+/**
+ * Generate registration options for creating a new passkey
+ */
+export async function createRegistrationOptions(
+	user: { id: string; email: string },
+	existingPasskeys?: Passkey[],
+): Promise<PublicKeyCredentialCreationOptionsJSON> {
+	// Get user's existing passkeys to exclude them from registration
+	const passkeys = existingPasskeys ?? (await db.select().from(table.passkey).where(eq(table.passkey.userId, user.id)));
+
+	const options = await generateRegistrationOptions({
+		rpName: env.RP_NAME,
+		rpID: env.RP_ID,
+		userName: user.email,
+		userDisplayName: user.email,
+		// Exclude existing credentials so user can't register the same key twice
+		excludeCredentials: passkeys.map((passkey) => ({
+			id: passkey.id,
+			type: "public-key",
+			transports: passkey.transports ?? [],
+		})),
+		authenticatorSelection: {
+			// Allow both platform (TouchID) and cross-platform (YubiKey) authenticators
+			authenticatorAttachment: undefined,
+			// Require discoverable credentials (resident keys) for usernameless login
+			residentKey: "required",
+			// Prefer user verification (biometric/PIN) but don't require for security keys
+			userVerification: "preferred",
+		},
+		// Support both modern and legacy authenticator algorithms
+		supportedAlgorithmIDs: [-7, -257],
+	});
+
+	return options;
+}
+
+/**
+ * Verify and store a passkey registration response
+ */
+export async function verifyAndStoreRegistration(
+	response: RegistrationResponseJSON,
+	expectedChallenge: string,
+	userId: string,
+	deviceName?: string,
+): Promise<VerifiedRegistrationResponse> {
+	const verification = await verifyRegistrationResponse({
+		response,
+		expectedChallenge,
+		expectedOrigin: env.RP_ORIGIN,
+		expectedRPID: env.RP_ID,
+		requireUserVerification: false, // Allow security keys without biometrics
+	});
+
+	if (!verification.verified || !verification.registrationInfo) {
+		throw new Error("Passkey registration verification failed");
+	}
+
+	const { credential, credentialBackedUp } = verification.registrationInfo;
+
+	// Store the passkey in the database
+	await db.insert(table.passkey).values({
+		id: credential.id,
+		userId,
+		publicKey: encodeBase64url(credential.publicKey),
+		counter: credential.counter,
+		deviceName: deviceName || "Unnamed device",
+		transports: credential.transports as ("usb" | "nfc" | "ble" | "internal" | "hybrid")[],
+		backedUp: credentialBackedUp,
+		lastUsedAt: new Date(),
+		createdAt: new Date(),
+		updatedAt: new Date(),
+	});
+
+	return verification;
+}
+
+/**
+ * Generate authentication options for signing in with a passkey
+ */
+export async function createAuthenticationOptions(userEmail: string): Promise<PublicKeyCredentialRequestOptionsJSON> {
+	// Look up user by email
+	const [user] = await db.select().from(table.user).where(eq(table.user.email, userEmail)).limit(1);
+
+	if (!user) {
+		// Don't reveal that user doesn't exist - return generic options
+		// This prevents email enumeration
```

Comment:
Assuming any reasonable percentage of our users use passkeys, it doesn't, since any real user will return passkeys while this won't.